sotodlib/io/imprinter.py

```diff
@@ -682,9 +686,132 @@ def delete_book_files(self, book):
             self.logger.warning(f"Failed to remove {book.path}: {e}")
             self.logger.error(traceback.format_exc())
 
+    def get_book_times(self, book):
+        """
+        Given a book of co-sampled data with a timing system, this function will
+        return a gap-filled list of timestamps that should be contained in the book.
+
+        Parameters
+        ----------
+        book: Book object
+        """
+        filedb = self.get_files_for_book(book)
+        t0 = book.start.timestamp()
+        t1 = book.stop.timestamp()
+        for i, files in enumerate(filedb.values()):
+            _files = sorted(files)
+            if i == 0:  # Get full list of timestamps for first file list
```

Comment:
A downside to using only one wafer to determine the list of timestamps is that the filled-in samples may not line up exactly with the samples that exist in other wafers. It probably doesn’t matter since we’ll flag those samples anyway, but I thought I should point that out.

I realize that checking all the wafers adds computational time, so maybe it could exist as a user-option to perform a more “comprehensive” search: the times for each wafer is assembled and trimmed, then look for unique times in the whole set. (Uniqueness only works when the timing system is on, but the correct fallback for no timing system is to make single-wafer books where it’s okay if samples don’t line up.) The vector of unique times is then filled with the missing times.

Reply:
Not sure I understand you here. Are you talking about if the slots are not actually co-sampled from a timing system?

It kind of seems like you are talking about a case where the filled timestamps are something like:
`[0, 1, 2, 3, 4, 5]`, but one sample happens to have an extra timestamp like `[0, 1, 2, 3, 3.5, 4, 5]`.
I think the correct action here would be to not include the `3.5` sample of the second array, which I think this method will actually correctly handle.

I'm probably misinterpreting you're concern though... do you have like an example of timestamp arrays that demonstrates your thought?

Reply:
I mean something like this: the filled-in timestamps are based on the mean/median dt, right? so if you have `[0, 1, 2, 3, 5]` in the first wafer, the algorithm would think the missing sample should be `4` when going with the median (or `4.25` when based on the mean). But it may turn out that all the other wafers actually had `[0, 1, 2, 3, 4.1, 5]`. The filled-in time value isn’t exactly what it should be. (This is a contrived example, but the idea is that your mean/median is only approximating where the true timestamps are)

But we’re going to be flagging those filled-in samples anyway, so maybe this doesn’t matter in the end.

Reply:
Right... I'm not exactly sure what the correct behavior would be in this case.... maybe @mhasself can comment, I'm sure this has been thought through for other projects.

As far as I'm aware, these issues where the timing sample is off by a large amount and it's not due to a dropped sample are exceedingly rare / not possible, so I'm not sure how much we want to concern ourselves with those edge cases. If we are concerned... I think the right thing to do is to first decide how we want to be handling those cases, write tests, and then afterwards modify the function so it handles them properly.

Reply:
The fill isn't based on median, actually ... it's a linterp, which will interpolate exactly between 3 and 5.  So you'll get 4.

However -- which of these new functions are designed to work on data where: (A) the timing system is in operation, and we can probably make multi-wafer books vs. (B) the timing system is not in operation.

The (A) case will be very stable... interpolating will bring only tiny errors, and I think that's fine.  Though you could pretty easily check+patch with the other vectors of timestreams, and this would be a great place to do it.

The (B) case can have very noisy timestamps... but of course in that case you don't need to worry about consistency between multiple wafers because you'll only have one wafer.

I don't think you need to have the "same code" run in cases (A) and (B).  It may be better to have separate, specialized functions for checking / filling / preparing the timestamp vectors in the two cases.